RFC: CKB Light Client Protocol (#370)
nervosnetwork/rfcs: rfcs/0000-ckb-light-client/0000-ckb-light-client.md

---
Number: "0000"
Category: Standards Track
Status: Proposal

> Author: Author: Boyu Yang <yangby@cryptape.com>

Reviewer: ```suggestion
Author: Boyu Yang <yangby@cryptape.com>
```

Created: 2022-08-18
---

# CKB Light Client Protocol

## Abstract

This RFC describes a light client protocol on CKB which allows clients to
verify that a blockchain is valid with limited resources.

## Motivation

Downloading and verifying all blocks is taking hours and requiring gigabytes

> of bandwidth and storage.  Hence, clients with limited resources cannot

Reviewer 2: `  Hence` two spacing here.

verify transactions independently without trusting full nodes.
Even some light clients only downloads  all block headers, the storage and
bandwidth requirements of those clients still increase linearly with the
chain length.
So CKB offers a solution base on [FlyClient] which is an efficient light
client for chains of variable difficulty. It requires downloading only a
logarithmic number of block headers while storing only a single block header
between executions.

## Background

### Merkle Mountain Range (MMR)

A [Merkle Mountain Range (MMR)][MMR] is a binary hash tree that allows for
efficient appends of new leaves without changing the value of existing
nodes.

In MMR, we use the insertion order to reference leaves and nodes.
We insert a new leaf to MMR as following steps:
- Insert leaf or node to next position.

> - If the current position has a left sibling, we merge the left and right

Reviewer: ```suggestion
- If the new inserted leaf or node has a left sibling, we merge the left and right
```

  nodes to produce a new parent node, then go back to step 1 to insert the
  node.

For example, we insert a leaf to the example MMR which has 11 leaves as
following steps:
- Insert leaf to next position: `19`.
- Then check the left sibling `18` and calculate parent node:
  `merge(mmr[18], mmr[19])`.
- Insert parent node to position `20`.
- Since the node `20` also has a left sibling `17`, calculate parent node:
  `merge(mmr[17], mmr[20])`.
- Insert new node to next position `21`.
- Since the node `21` have no left sibling, complete the insertion.

```
# An MMR with 11 leaves:
          14
       /       \
     6          13
   /   \       /   \
  2     5     9     12     17
 / \   /  \  / \   /  \   /  \
0   1 3   4 7   8 10  11 15  16 18

# After insertion of a new leaf:

          14
       /       \
     6          13            21
   /   \       /   \         /   \
  2     5     9     12     17     20
 / \   /  \  / \   /  \   /  \   /  \
0   1 3   4 7   8 10  11 15  16 18  19
```

#### Merkle Root

> An MMR is constructed by one or more sub merkle trees (or mountains). Each

Reviewer: It's better to give a formal definition of how to *bag* the peaks. For the example, it's hard to deduce how to compute the root containing 4 peaks.

sub merkle tree's root is a peak in MMR, we calculate the MMR root by
bagging these peaks from left to right.

For example:
- In the above 11 leaf MMR we have 3 peaks: `14, 17, 18`, we bag these peaks
  from left to right to get the root:
  `merge(merge(mmr[14], mmr[17]), mmr[18])`.
- In the above 12 leaf MMR we have 2 peaks: `14, 21`, we bag these peaks
  from left to right to get the root: `merge(mmr[14], mmr[21])`.

#### Merkle Proof

The merkle proof is an array of hashes constructed with the following parts:
- A merkle proof from the leaf's sibling to the peak that contains the leaf.
- A hash that bags all right-hand side peaks, skip this part if no
  right-hand peaks.
- Hashes of all left-hand peaks from right to left, skip this part if no
  left-hand peaks.

We can reconstruct the merkle root from the proofs.

> ### The FlyClient Sampling Protocol Under Variable Difficulty

Reviewer: This section assumes that readers have a good understanding of the Flyclient paper. It omits too much explanation of terms and background.

The client uses $g(x) = \frac{1}{(1-x)\ln{\delta}}$ as the probability
density function (PDF) to sample blocks.

> In this probability, $x$ denotes the relative aggregate difficulty weight

Reviewer: This paragraph introduces two terms without any explanation.

Reviewer 3: Which terms?
Did you mean "the relative aggregate difficulty weight" and "the relative difficulty weight"? :cry:
These two terms just like "a big fat black cat". 
In the original paper, just these adjectives are their whole explanations.

sampled with probability $1$.

Concretely, let $\delta = c ^ {k}$, where $c$ denote the fraction of the
adversary's computing power relative to the honest computing power, $p$
denote the probability of catching the adversary with a single sample and
$k = \frac{1}{p}$.

The sampling domain is restricted from $0$ to $1 − \delta$ and the blocks in
the last $\delta$ fraction region should always be checked directly.

Let $n$ denote the number of the blocks in the region which requires
verifying and $L$ denote the number of blocks in the last $\delta$ fraction
region. Obviously, we have $L = \delta \times n = c ^ {k} \times n$.
Accordingly, $k = \log_{c}{(\frac{L}{n})}$.

Define $p_{m} = (1 - \frac{1}{k})^{m}$ as the probability of failure, i.e.,
not catching the optimal adversary after check $m$ blocks. If we want
$p_{m} \le 2^{-\lambda}$, then
$m \ge \frac{\lambda}{\log_{\frac{1}{2}} {(1 - \frac{1}{k})}}$.

## Specification

### MMR Node Specification

We treat all blocks as the leaf nodes of an MMR, and its block number is the
index of leaves in that MMR.

#### MMR Node

Each MMR node is defined as follows:

- `hash`

>   - For a leaf node, it's an empty hash (`0x0000...0000`).

Reviewer: If all the leaves are empty hashes, why bother creating the tree? Given a specified number of leaves, the tree is deterministic.

Reviewer 3: This is correct, but the description for non-leaf nodes is confused, I will update it.

How about the description as following?

```
A hash for a node is the hash of its concatenate serialized data.
- For a leaf node, its "hash" field is an empty hash (`0x0000...0000`), since it has no children.
- For a non-leaf node, its "hash" field the hash of its child nodes' hashes (concatenate serialized data of two hashes).
```

Reviewer: I think the confusion comes from the bad field name `hash`. A better name could be `children_hash`.

Reviewer 3: Updated!

    transaction hashes,
  - the proofs that proving the those valid transactions are in those blocks,
  - hashes of transactions which are from the provided transaction hashes
    but not in the current chain.

  Otherwise, the server should only return its current verifiable tip
  header.

## Limitations

To avoid attacks by malicious clients, there are few limitations in server
side.
- In `GetLastStateProof` messages, the sum of the size of difficulties and 2
  times of the `last_n_blocks` should NOT greater than 1000.
- In `GetBlocksProof` messages, the size of block hashes should NOT greater
  than 1000.
- In `GetTransactionsProof` messages, the size of transactions hashes should
  NOT greater than 1000.

## Deployment

Since CKB doesn't record the hashes of chain roots into headers at the start,
so we use an activation epoch number to enable this feature [\[2\]].

And we also have to create all MMR nodes for blocks before that epoch to
make sure that the indexes of the MMR leaves are equal to the block numbers.

If users have data which generated by the previous version of CKB, they have
to do data migration before enable this feature.

> And once the migration started, the data will be no longer compatible with

Reviewer 2: ```suggestion
And once the migration started, the data will no longer be compatible with
```
